HKDF 1: PSA: implement HKDF_Expand and HKDF_Extract algorithms

mprse opened the pull request:
## Description
Implement HKDF-Expand and HKDF-Extract as separate algorithms in the PSA API. 

Resolves https://github.com/Mbed-TLS/mbedtls/issues/5784


## Status
**READY**

## Requires Backporting
 NO

mpg commented on library/psa_crypto.c:
```diff
@@ -4411,40 +4414,49 @@ static psa_status_t psa_key_derivation_hkdf_read( psa_hkdf_key_derivation_t *hkd
             return( PSA_ERROR_BAD_STATE );
 
+
+        if( PSA_ALG_IS_HKDF_EXTRACT( kdf_alg ) && hkdf->block_number == 0 )
```
Ok, I think this is correct, and the rest of the changes in this function as well, but I find this function pretty hard to reason about it is current state, because it's just doing very different things depending on whether the alg is HKDF-Extract on one hand, or HKDF / HKDF-Expand on the other hand. So different in fact, that I think it would be better to just have two separate functions here, and dispatch to the right one in `psa_key_derivation_output_bytes()`. I think the function for HKDF-Extract would be very simple as it can copy directly from `prk` and doesn't need to care about the case the `ouput_length` is more than one block.

It's all too easy when adding new features to existing code to slowly increase the complexity and decrease the readability, so let's pay extra attention to that!

mprse replied:
I agree that it became more complex, but I think this is the easiest way to handle reading key in parts.
To be honest I don't understand why we allow reading key in parts, but it works like this for other algorithms so it should also work like this for reading `prk` (`HKDF-Extract`).
I can add function `psa_key_derivation_hkdf_extract_read()`, but the loop that allows reading parts of key will be duplicated.

mpg replied:
Thanks for trying it out! Agreed, using the same field, in this case `hkdf->offset_in_block`, for two different things doesn't sound great either. I'll think about it.

> To be honest I don't understand why we allow reading key in parts, but it works like this for other algorithms so it should also work like this for reading prk (HKDF-Extract).

Yes, for other KDFs it is useful to be able to read the output in parts, for example TLS 1.2 uses its PRF by drawing n bits for the encryption key, n bits for the decryption key, etc. For HKDF-Extract I don't think it ever makes sense to read the output in parts, but it should probably work like other algorithms for the sake of uniformity.

mpg replied:
Ok, I tried writing an alternative proposal for how to handle this, you can find it here: the [branch](https://github.com/mpg/mbedtls/tree/5834-prop), the [commit](https://github.com/mpg/mbedtls/commit/3676419d5813d417ad216a385bdb1c773a0436cc). Compared to your version, it does more in `hkdf_input()` and less in `hkdf_read()`. IMO that reflects the difference between Extract and the other two: with Extract the output is pre-computed when reading the key, while with HDKF(-Expand) it's computed on-demand when reading it.

It's of course debatable, and partially a matter of taste, which version is better. But I think I can identify two specific points that bugged me when first reading the version in the present PR:

1. The `memcpy()` from `prk` to `output_block` is visually inside a while loop, but only happens once. IMO it should ideally not look like it's in a loop, or at least there should be a comment explaining the flow (the loop will run twice: the first time generating no output, then doing the memcpy, then it will generate output then stop because we know we're not being asked for more than one block of output).
2. The limit we check is always 0xff even for Extract. This is not just readability: if we have a defensive check it should work for all cases, but I think for me it impacted readability as well, making it harder to understand how many times the loop will run.

I'm not necessarily pushing for you to adopt my proposal, but whatever solution you come up with, I'd like it to address the above points.

mpg replied:
(Posted the comment above just before I saw you pushed new commits, will have a look at them now.)

mprse replied:
Reviewed the proposal and it looks good to me. I like the idea to copy `prk` to `output` immediately after it was generated. This simplifies `hkdf_read()`. I will adapt code.
Thanks!

mpg replied:
Cool, I'm happy we reached a quick agreement on that one!

mpg commented on library/psa_crypto.c:
```diff
@@ -5177,32 +5195,48 @@ static psa_status_t psa_hkdf_input( psa_hkdf_key_derivation_t *hkdf,
         case PSA_KEY_DERIVATION_INPUT_SECRET:
             /* If no salt was provided, use an empty salt. */
-            if( hkdf->state == HKDF_STATE_INIT )
+            if( PSA_ALG_IS_HKDF_EXPAND( kdf_alg ) )
```
I think the comment above is misplaced now, and should go in the `else` branch, above `if( hkdf->state == HKDF_STATE_INIT )`.

mpg reviewed (changes requested):
Looking pretty good. I like that you shared code as much as possible between the three HKDF variants, this is good for code size, and I think it works well for `hkdf_input()`, but for `hkdf_read()` IMO the way it's done impacts readability so I'm suggesting another approach.

Thanks for including not just test vectors from the RFC but also negative tests. According to lcov, there are a few cases that are not covered yet that we should try covering if we can (or comment why they can't be reach through the public API, like the ` hkdf->block_number == 0xff` check in `hkdf_read()`):
- `hkdf->state != HKDF_STATE_INIT` in `hkdf_input()` for `SECRET` with `HKDF_EXPAND`;
- `data_length > sizeof( hkdf->prk )` in the same case.
Other than those two, branch coverage is looking great, good job!

mprse commented:
>According to lcov, there are a few cases that are not covered yet that we should try covering if we can (or comment why they can't be reach through the public API, like the hkdf->block_number == 0xff check in hkdf_read()):
hkdf->state != HKDF_STATE_INIT in hkdf_input() for SECRET with HKDF_EXPAND;
data_length > sizeof( hkdf->prk ) in the same case.

Are the icov results available on CI? 


- `hkdf->block_number == 0xff`: is defensive code to check if the object is corrupted, so it can't be covered in normal operation.
- `hkdf->state != HKDF_STATE_INIT` in `hkdf_input()` for `SECRET` with `HKDF_EXPAND`: it is also defensive code that should not happen in normal operation as for `HKDF_EXPAND` we alow only two inputs: `SECRET` and `INFO` and `INFO` does not modify `hkdf->state`, so this also can happen only if `hkdf` object is corrupted.
- `data_length > sizeof( hkdf->prk )`: this should be covered, but maybe this condition should be changed not to check against buffer size, but expected `prk` szie (`data_length != PSA_HASH_LENGTH( hash_alg )`)?

mpg commented:
> Are the icov results available on CI?
> 
No, I don't think so, actually I don't think we even run coverage for each PR right now, though we have plans to do so in the future. I generated the data locally with:
```
make CFLAGS='--coverage -g3 -Og' LDFLAGS='--coverage' -C tests test_suite_psa_crypto 
(cd tests && ./test_suite_psa_crypto)
make lcov
```
Then the results are in `Coverage/` and can be viewed with your usual browser.

>     * `hkdf->block_number == 0xff`: is defensive code to check if the object is corrupted, so it can't be covered in normal operation.

Agree, I was precisely citing that as an example where it can't be covered and there's a comment in the source explaining why, so it's perfectly OK.

>     * `hkdf->state != HKDF_STATE_INIT` in `hkdf_input()` for `SECRET` with `HKDF_EXPAND`: it is also defensive code that should not happen in normal operation as for `HKDF_EXPAND` we alow only two inputs: `SECRET` and `INFO` and `INFO` does not modify `hkdf->state`, so this also can happen only if `hkdf` object is corrupted.

Ok. Perhaps add a comment in the code then, like "this can only happen if the hkdf object is corrupted".

>     * `data_length > sizeof( hkdf->prk )`: this should be covered, but maybe this condition should be changed not to check against buffer size, but expected `prk` szie (`data_length != PSA_HASH_LENGTH( hash_alg )`)?

Good point! Let's change that and then test it.

mpg reviewed (commented):
Thanks for addressing my feedback. Looks good to me, except for the pending point about how to optimize readability.

mprse commented:
Thanks for information how to generate coverage report! I tried it on my side and it works but the results are strange:
![image](https://user-images.githubusercontent.com/30721012/169253991-f6371a44-5b84-43f0-bbee-41e9628bd371.png)

- `if( hkdf->state != HKDF_STATE_INIT )` is fully covered and according to my understanding this condition should be always false.
- `if( data_length != PSA_HASH_LENGTH( hash_alg ) )` is now also fully covered but we have only one test case when this condition is true, but in the results there are 8 pluses.

Also how it it possible that first condition has one plus and one minus and second condition a lot more?

mpg reviewed (approved):
LGTM!

mpg commented:
> Also how it it possible that first condition has one plus and one minus and second condition a lot more?

I think that's because lcov counts the branches in the expansion of macros, and `PSA_HASH_LENGTH()` expands to a series of ternary operators, so a vast number of branches. In this case it's unfortunate branches inside macros are shown, as it's just noise here, but in other cases it could be useful.

Regarding line 5203, it shows "branch 0" to be taken (blue "+") (if you hover on it it will tell you "30 times") and "branch 1" to be never taken (red "-"). I don't know how branches are numbered but in that case I assume branch 0 is the condition was false (ie, not returning an error here) while branch 1 is the condition was true. So, based on those assumptions I think the results are consistent with what we expect.

On thing I find very odd is that line 5204 has a white background, as if is wasn't an executable statement. I think it should be red to clearly show that we never returned early from here. That's not the first time I see this kind of thing with lcov, and I'm not sure what the issue is. Seems to affect all conditional `return` statements here...

Anyway, I sure wish lcov's output was clearer, but in that case it seems at least compatible with what we expect.

mprse commented:
> Regarding line 5203, it shows "branch 0" to be taken (blue "+") (if you hover on it it will tell you "30 times") and "branch 1" to be never taken (red "-"). I don't know how branches are numbered but in that case I assume branch 0 is the condition was false (ie, not returning an error here) while branch 1 is the condition was true. So, based on those assumptions I think the results are consistent with what we expect.

Indeed when you hover on "+"/"-" it gives more details, but it is really strange that 5204 is not marked red. Seems that this tool is not perfect. Anyway thanks for short test coverage training!

mprse commented:
It would be good to move forward with this one as other PRs depend on it: 
- https://github.com/Mbed-TLS/mbedtls/pull/5838
- https://github.com/Mbed-TLS/mbedtls/pull/5867 (2 issues + other follow ups)

CI was green, but now has been invalidated. Not sure if force-push to trigger CI is a good option as @mpg review will be invalidated.

cc @gilles-peskine-arm

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -4403,14 +4406,15 @@ static psa_status_t psa_key_derivation_hkdf_read( psa_hkdf_key_derivation_t *hkd
+        if( hkdf->block_number == last_block )
             return( PSA_ERROR_BAD_STATE );
 
+
```
Two newlines

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -4425,24 +4429,24 @@ static psa_status_t psa_key_derivation_hkdf_read( psa_hkdf_key_derivation_t *hkd
             status = psa_mac_update( &hkdf->hmac,
-                                     hkdf->output_block,
-                                     hash_length );
+                                    hkdf->output_block,
```
Formatting not in line here and below

mprse replied:
Fixed

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -5488,10 +5531,10 @@ static psa_status_t psa_key_derivation_input_internal(
 
 #if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
-    if( PSA_ALG_IS_HKDF( kdf_alg ) )
+    if( PSA_ALG_IS_HKDF( kdf_alg ) || PSA_ALG_IS_HKDF_EXTRACT( kdf_alg ) ||
```
Four places use the same set of conditions - maybe consider refactoring? Minor suggestion.

mprse replied:
Fixed

AndrzejKurek reviewed (changes requested):
Only minor requests. Good job!
The code is a little bit less readable now, but I think it's a good tradeoff considering code reusability. 
Also - I would say that it screams for OOP (at least from my perspective) :D

mprse commented on include/psa/crypto_values.h:
```diff
@@ -1762,6 +1762,97 @@
+ * HKDF-Expand using HMAC-SHA-256.
+ *
+ * This key derivation algorithm uses the following inputs:
+ *  - PSA_KEY_DERIVATION_INPUT_SECRET is the pseudoramdom key (PRK).
```
Fixed

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -5117,6 +5117,294 @@ PSA key derivation: HKDF SHA-1, RFC5869 #7, output 42+0
+depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256
+derive_output:PSA_ALG_HKDF_EXTRACT(PSA_ALG_SHA_256):PSA_KEY_DERIVATION_INPUT_SALT:"000102030405060708090a0b0c":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SECRET:"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":PSA_SUCCESS:0:"":PSA_SUCCESS:0:"":PSA_SUCCESS:"":32:"077709362c2e32df0ddc3f0dc47bba6390b6c73bb50f9c3122ec844ad7c2b3":"":0:0:0
+
+PSA key derivation: HKDF-Extract SHA-256, RFC5869 #1, out 1+30 b
```
```suggestion
PSA key derivation: HKDF-Extract SHA-256, RFC5869 #1, out 1+31 b
```

mprse replied:
Adapted expected output as `1+31` case already exists.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -5460,6 +5756,22 @@ PSA key derivation: HKDF SHA-1, request too much capacity
+PSA key derivation: HKDF-Expand SHA-1, request too much capacity
+depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_1
+derive_set_capacity:PSA_ALG_HKDF_EXPAND(PSA_ALG_SHA_1):255 * PSA_HASH_LENGTH(PSA_ALG_SHA_1) + 1:PSA_ERROR_INVALID_ARGUMENT
+
```
Test coverage: good job, except that there's one class of negative tests that's missing: a bad input sequence (incorrect step, duplicated step, missing step, step out of order).

It's annoying both to write and review such test cases. There are such tests for the existing KDF algorithms, but they're incomplete. So I propose to omit them here, and instead work on a test case generator that will enumerate bad sequences of inputs.

mprse replied:
Ok, so I think we need a task for this?

mprse commented on library/psa_crypto.c:
```diff
@@ -4295,7 +4295,8 @@ psa_status_t psa_key_derivation_abort( psa_key_derivation_operation_t *operation
 #if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
-    if( PSA_ALG_IS_HKDF( kdf_alg ) )
+    if( PSA_ALG_IS_HKDF( kdf_alg ) || PSA_ALG_IS_HKDF_EXTRACT( kdf_alg ) ||
+        PSA_ALG_IS_HKDF_EXPAND( kdf_alg ) )
```
Added `PSA_ALG_IS_ANY_HKDF` :  https://github.com/Mbed-TLS/mbedtls/pull/5834/commits/baa610fd0f072e4b9c128afcce5619bc83ac3a11

mprse commented on tests/scripts/generate_psa_tests.py:
```diff
@@ -78,6 +78,11 @@ def automatic_dependencies(*expressions: str) -> List[str]:
     """
     used = set()
     for expr in expressions:
+        # HKDF_EXTRACT and HKDF_EXPAND algs depend on HKDF
```
I can see that `PSA_WANT_ALG_HKDF` is used as dependency in test and  `#if` guards use `MBEDTLS_PSA_BUILTIN_ALG_HKDF`. Do I need to add also `MBEDTLS_PSA_BUILTIN_ALG_HKDF_EXTRACT/EXPAND` and add guards for related code?

gilles-peskine-arm replied:
Yes please. We should treat all algorithms the same way unless there's a very good reason: for us it simplifies reasoning about the code and testing, and for users it simplifies the understanding of how to use the library (which also reduces our support load).

mprse replied:
I understand that both should be still defined under `MBEDTLS_HKDF_C`:

```
#if defined(MBEDTLS_HKDF_C)
#define MBEDTLS_PSA_BUILTIN_ALG_HMAC 1
#define PSA_WANT_ALG_HMAC 1
#define MBEDTLS_PSA_BUILTIN_ALG_HKDF 1
+#define MBEDTLS_PSA_BUILTIN_ALG_HKDF_EXTRACT 1
+#define MBEDTLS_PSA_BUILTIN_ALG_HKDF_EXPAND 1
#define PSA_WANT_ALG_HKDF 1
+#define PSA_WANT_ALG_HKDF_EXTRACT 1
+#define PSA_WANT_ALG_HKDF_EXPAND 1
#endif /* MBEDTLS_HKDF_C */
```

or we need also `MBEDTLS_HKDF_EXTRACT_C` and `MBEDTLS_HKDF_EXPAND_C`?

gilles-peskine-arm replied:
Actually the PSA HKDF implementation does not use `hkdf.c`. So that part of `include/mbedtls/config_psa.h` should not have been there at all! The PSA HKDF can be available whenever the underlying hash is. https://github.com/Mbed-TLS/mbedtls/issues/5890 (please don't fix it here, but don't add HKDF_EXxxx there either)

mprse commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -7102,6 +7109,10 @@ void derive_output( int alg_arg,
     if( derive_type == 1 ) // output key
     {
         psa_status_t expected_status = PSA_ERROR_NOT_PERMITTED;
+        size_t bits = 48; // default for Mix-PSK-to-MS
```
This part was created for Mix-PSK-to-MS to test output key and for Mix-PSK-to-MS output key is always 48 bits long.
Now it is also used to test HKDF-Extract/Expand and for these output key length is not constant.

How to make it more generic?

mprse replied:
Oh I mixed things, sorry. 
Thanks for pointing this.

mprse commented:
I addressed all the review comments, but probably now CI won't be happy.
Will continue on Monday.

mprse commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -5117,6 +5117,294 @@ PSA key derivation: HKDF SHA-1, RFC5869 #7, output 42+0
+depends_on:PSA_WANT_ALG_HKDF:PSA_WANT_ALG_SHA_256
+derive_output:PSA_ALG_HKDF_EXTRACT(PSA_ALG_SHA_256):PSA_KEY_DERIVATION_INPUT_SALT:"000102030405060708090a0b0c":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SECRET:"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b":PSA_SUCCESS:0:"":PSA_SUCCESS:0:"":PSA_SUCCESS:"":32:"077709362c2e32df0ddc3f0dc47bba6390b6c73bb50f9c3122ec844ad7c2b3":"":0:1:0
+
+PSA key derivation: HKDF-Extract SHA-256, RFC5869 #1, out 1+30 k
```
Adapted expected output as `1+31` case already exists.

mprse commented:
I'm trying to push last changes through CI and encountered a strange issue and got stuck. 

For example `component_test_psa_crypto_config_accel_cipher` component is failing as follows:
```
test_suite_psa_crypto:
PSA MAC sign: RFC4231 Test case 1 - HMAC-SHA-224 .................. FAILED
PSA MAC verify: RFC4231 Test case 1 - HMAC-SHA-224 ................ FAILED
PSA MAC sign: RFC4231 Test case 1 - HMAC-SHA-384 .................. FAILED
PSA MAC verify: RFC4231 Test case 1 - HMAC-SHA-384 ................ FAILED
PSA MAC sign: RFC4231 Test case 1 - HMAC-SHA-512 .................. FAILED
PSA MAC verify: RFC4231 Test case 1 - HMAC-SHA-512 ................ FAILED
PSA MAC verify: RFC4231 Test case 2 - HMAC-SHA-224 ................ FAILED
PSA MAC verify: RFC4231 Test case 2 - HMAC-SHA-384 ................ FAILED
PSA MAC verify: RFC4231 Test case 2 - HMAC-SHA-512 ................ FAILED
PSA MAC verify: RFC4231 Test case 3 - HMAC-SHA-224 ................ FAILED
PSA MAC verify: RFC4231 Test case 3 - HMAC-SHA-384 ................ FAILED
PSA MAC verify: RFC4231 Test case 3 - HMAC-SHA-512 ................ FAILED
PSA MAC verify: RFC4231 Test case 4 - HMAC-SHA-224 ................ FAILED
PSA MAC verify: RFC4231 Test case 4 - HMAC-SHA-384 ................ FAILED
PSA MAC verify: RFC4231 Test case 4 - HMAC-SHA-512 ................ FAILED
PSA MAC sign: HMAC-SHA-224, truncated to 28 bytes (actual size) ... FAILED
PSA MAC verify: HMAC-SHA-224, truncated to 28 bytes (actual size) . FAILED
PSA MAC sign: HMAC-SHA-512, truncated to 64 bytes (actual size) ... FAILED
PSA MAC verify: HMAC-SHA-512, truncated to 64 bytes (actual size) . FAILED
PSA MAC sign: HMAC-SHA-224, truncated to 27 bytes ................. FAILED
PSA MAC verify: HMAC-SHA-224, truncated to 27 bytes ............... FAILED
PSA MAC sign: HMAC-SHA-512, truncated to 63 bytes ................. FAILED
PSA MAC verify: HMAC-SHA-512, truncated to 63 bytes ............... FAILED
PSA MAC sign: HMAC-SHA-224, truncated to 4 bytes .................. FAILED
PSA MAC verify: HMAC-SHA-224, truncated to 4 bytes ................ FAILED
PSA MAC sign: HMAC-SHA-512, truncated to 4 bytes .................. FAILED
PSA MAC verify: HMAC-SHA-512, truncated to 4 bytes ................ FAILED
PSA key derivation: HKDF-SHA-512, good case ....................... FAILED
PSA key derivation: HKDF SHA-1, RFC5869 #4, output 42+0 ........... FAILED
PSA key derivation: HKDF SHA-1, RFC5869 #5, output 82+0 ........... FAILED
PSA key derivation: HKDF SHA-1, RFC5869 #6, output 42+0 ........... FAILED
PSA key derivation: HKDF SHA-1, RFC5869 #7, output 42+0 ........... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #4, out 42+0 k ..... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #5, out 82+0 k ..... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #6, out 42+0 k ..... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #7, out 42+0 k ..... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #4, out 42+0 b ..... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #5, out 82+0 b ..... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #6, out 42+0 b ..... FAILED
PSA key derivation: HKDF-Expand SHA-1, RFC5869 #7, out 42+0 b ..... FAILED
PSA key derivation: TLS 1.2 PRF SHA-384, output 148+0 ............. FAILED
PSA key derivation: TLS 1.2 PRF SHA-384, output 147+1 ............. FAILED
PSA key derivation: TLS 1.2 PRF SHA-384, output 1+147 ............. FAILED
PSA key derivation: TLS 1.2 PRF SHA-384, output 74+74 ............. FAILED
PSA key derivation: TLS 1.2 PRF SHA-384, output 74+73 ............. FAILED
PSA key derivation: TLS 1.2 PSK-to-MS, SHA-384, 48+0 .............. FAILED
PSA key derivation: TLS 1.2 PSK-to-MS, SHA-384, 24+24 ............. FAILED
PSA key derivation: TLS 1.2 PSK-to-MS, SHA-384, 0+48 .............. FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 48+0, ol 0 b .. FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 24+24, ol 0 b . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 0+48, ol 0 b .. FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 48+0, ol 20 b . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 24+24, ol 20 b  FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 0+48, ol 20 b . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 48+0, ol 48 b . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 24+24, ol 48 b  FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 0+48, ol 48 b . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 48+0, ol 20 k . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 24+24, ol 20 k  FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 0+48, ol 20 k . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 48+0, ol 48 k . FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 24+24, ol 48 k  FAILED
PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-384, 0+48, ol 48 k . FAILED
PSA key derivation: HKDF SHA-1, request maximum capacity .......... FAILED
PSA key derivation: HKDF-Expand SHA-1, request maximum capacity ... FAILED
PSA key derivation: largest possible key .......................... FAILED
FAILED (1538 / 1604 tests (95 skipped))
```

Checked few cases and it seems that the problem is common.  `psa_mac_sign_setup()/psa_mac_verify_setup` are failing because `operation` object is corrupted:
```
PSA MAC sign: RFC4231 Test case 1 - HMAC-SHA-224 .................. FAILED
  ( psa_mac_sign_setup( &operation, key, alg ) ) == PSA_SUCCESS
  at step 1, line 3226, test_suite_psa_crypto.c
  lhs = 0xffffffffffffff77 = -137
  rhs = 0x0000000000000000 = 0
```

![image](https://user-images.githubusercontent.com/30721012/172175854-bd9f2a7a-323f-4f41-9178-6046574715b2.png)

For example `mac_sign` test calls `psa_mac_sign_setup()` in loop and first iteration is ok, but second fails:
https://github.com/Mbed-TLS/mbedtls/blob/development/tests/suites/test_suite_psa_crypto.function#L2810

According to the documentation operation object can be reused after abort, so the test code is fine. When I tried to clear operation before `psa_mac_sign_setup()` the test was passing (I know that this is not the solution). For me it looks like something is writing to incorrect address. Tried to set set brake-points on copy operations added by this PR, but non of them was executed during MAC sign test.
Before I addressed review comments CI was green, but I don't see anything related to such failure in the last changes.

mprse commented:
I couldn't find relation between the changes and failing driver tests. I suspected memory corruption, but it turned out that that `MBEDTLS_PSA_ACCEL_ALG_HKDF_EXTRACT/EXPAND` symbols were missing.

mprse commented:
CI is green. Setting to r4r.

mpg reviewed (commented):
Looks all good to me except for one thing I don't understand.

mprse commented on tests/suites/test_suite_psa_crypto_storage_format.function:
```diff
@@ -151,6 +151,11 @@ static int can_exercise( const psa_key_attributes_t *attributes )
         PSA_ALG_NONE;
     psa_key_usage_t usage = psa_get_key_usage_flags( attributes );
 
+    /* Disable KA(ECDH,HKDF_EXTRACT/EXPAND...) test cases. */
```
When we added new new algs the storage format test generator generated test for those and and `KA(ECDH,HKDF_EXTRACT/EXPAND...)` are failing.  There was discussion on slack in the thread about this PR (yesterday) with @gilles-peskine-arm.
I disabled them at least for now to check if CI is now happy. We need to do something with them in this PR or in fallow up task. Let's wait for @gilles-peskine-arm review.

gilles-peskine-arm replied:
> I don't know how to handle these exceptions other way then skipping them in can_exercise().

Yes, and then write a manual test for that (or use an existing test function, I haven't checked). The manual test doesn't have to use a persistent key and doesn't have to try all possible curves and hashes, just a couple of cases (at least one where the KA output is too small, one where it's just right and one where it's too large).

gilles-peskine-arm reviewed (changes requested):
Now we aren't testing KA+HKDF_EXxxx enough. Other than that, looks good to me.

mprse commented on library/psa_crypto.c:
```diff
@@ -5221,7 +5314,9 @@ static psa_status_t psa_hkdf_input( psa_hkdf_key_derivation_t *hkdf,
     }
 }
-#endif /* MBEDTLS_PSA_BUILTIN_ALG_HKDF */
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_HKDF ||
```
Adding internal macro for any HKDF alg seems to me the optimal solution to reduce `ifdefs` and keep `PSA_WANT_ALG_HKDF_EXxxx`.

mprse commented:
- Added special macro to limit number of `#ifs` lines.
- Added special handling for `KA(ECHD,HKDF_EXTRACT/EXPAND)` tests.

One thing that still is missing are the following macros:
```
#define PSA_WANT_ALG_HKDF_EXTRACT 1
#define MBEDTLS_PSA_BUILTIN_ALG_HKDF_EXTRACT 1
#define PSA_WANT_ALG_HKDF_EXPAND 1
#define MBEDTLS_PSA_BUILTIN_ALG_HKDF_EXPAND 1
```

They shouldn't be declared under `MBEDTLS_HKDF_C`, but I can't find suitable place for them in `psa_config.h`.

mprse commented:
I looks like adding internal macro won't be so easy:
```
    > library/psa_crypto.c:93: 'MBEDTLS_PSA_BUILTIN_ALG_ANY_HKDF' looks like a
      typo. It was not found in any macros or any enums. If this is not a typo,
      put //no-check-names after it.
      |
   93 | #define MBEDTLS_PSA_BUILTIN_ALG_ANY_HKDF 1
      | 
    > library/psa_crypto.c:4243: 'MBEDTLS_PSA_BUILTIN_ALG_ANY_HKDF' looks like a
      typo. It was not found in any macros or any enums. If this is not a typo,
      put //no-check-names after it.
      |
 4243 | #if defined(MBEDTLS_PSA_BUILTIN_ALG_ANY_HKDF) || \
      |
...
```

Should I add `//no-check-names` to definition and all occurrences?

mpg commented:
I think the problem is that `check_names.py` expects all macros to either be defined in a header file (public or internal), or have a name that doesn't start with `MBED` - you can see it has no issue with `RNG_NOT_INITIALIZED` defined a few lines below in the file. Probably the easiest way to pacify it is to meet one of those expectations. Adding `// no-check-names` everywhere doesn't sound very appealing.

mprse commented:
@gilles-peskine-arm I added exception for the storage format Extract/Expand tests. To do that I added additional parameters to test function. I wanted to use existing code and not duplicate code a specially the test is complicated. Unfortunately CI is unhappy now:
```
[2022-06-10T18:44:58.546Z] Test case from tests/suites/test_suite_psa_crypto_storage_format.v0.data line 21 "read lifetime: PERSISTENT" has disappeared: key_storage_read:PSA_KEY_LIFETIME_PERSISTENT:PSA_KEY_TYPE_RAW_DATA:8:PSA_KEY_USAGE_EXPORT:0x0000:0x0000:"4c":"505341004b455900000000000100000001100800010000000000000000000000010000004c":0

[2022-06-10T18:44:58.546Z] FAIL: 1928/3782 storage format test cases have changed or disappeared.

[2022-06-10T18:44:58.546Z] Info: number of storage format tests cases: 3782 -> 4114.

script returned exit code 1
```

mpg reviewed (commented):
Just to record that I have reviewed until here. Just need to resolve the issue with testing, other than that I'm happy.

mprse commented:
I'm thinking now to remove additional parameters and add additional test flags to handle special cases.

gilles-peskine-arm commented on tests/include/test/psa_exercise_key.h:
```diff
@@ -242,15 +242,23 @@ int mbedtls_test_psa_exported_key_sanity_check(
 int mbedtls_test_psa_exercise_key( mbedtls_svc_key_id_t key,
                                    psa_key_usage_t usage,
-                                   psa_algorithm_t alg );
+                                   psa_algorithm_t alg,
```
I don't like adding all this complexity, just for the sake of a few unusual cases with key derivation. This would be better handled with a dedicated function for key derivation.

gilles-peskine-arm reviewed (changes requested):
Please revert d17541acc1431548da3bf03aa2b203142f6a8451 and 5e123137f8cdb711907d15f7b9ef71b05dea90a0 or rewrite the history to remove them. This approach is too complicated and isn't actually doing what we want (HKDF_EXTRACT should be exerciseable normally).

Instead, please just add an exception to `can_exercise` for HKDF_EXPAND when the secret size isn't the hash size. And for those cases, write (or reuse) a separate test function, not in the storage format tests but in `test_suite_psa_crypto`.

gilles-peskine-arm commented:
> Just need to resolve the issue with testing, other than that I'm happy.

The only failure is the storage format tests reported by the ABI check job, and at first glance they are acceptable changes (I haven't checked in detail because I disagree with the approach — I think those changes are not necessary). The ABI check job notices that some storage format test cases have disappeared. It isn't smart enough to realize that there are now very similar test cases that just have extra parameters. We would need to check manually that the new test cases do indeed supersede all the old ones.
